---
title: "Analysis using GSEA for R"
author: "Jennifer Crockett"
date: "2024-03-08"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 10, fig.height = 10)
library(tidyverse)
library(here)
library(scico)
library(lemon)
devtools::load_all("/Users/jcrockett/Documents/jobs/code_sample_projects/GSEAforR/src/R")

theme_set(theme_light(base_size = 14))
```

# Analysis description

DNMT3A is a gene that is commonly mutated in acute myeloid leukemia (AML), a type of bone marrow/blood cancer. Researchers want to know why this gene mutation causes AML. A group of researchers introduced an AML-associated DNMT3A mutation into the bone marrow and blood cells of mice so that they could study the changes caused by this mutation, compared to normal mice. They collected specialized bone marrow/blood cells from the mice and performed RNA-seq gene expression analysis. Their dataset was deposited at a public gene expression database: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72737

In this analysis, the _GSEA for R_ functions will be used to perform GSEA analysis on data from DNMT3A mutant vs. DNMT3A non-mutant (wild type, or 'WT') cells.

## Dataset reference

Guryanova OA, Shank K, Spitzer B, Luciani L et al. DNMT3A mutations promote anthracycline resistance in acute myeloid leukemia via impaired nucleosome remodeling. Nat Med 2016 Dec;22(12):1488-1495. PMID: 27841873

# Load data

Load the preprocessed data generated in `GSEAforR/testing/data/Test_data_preprocessing.Rmd`. This consists of gene expression data and sample metadata.

```{r}
expr <- read_tsv(here("testing", "data", "GSE72737_preprocessed_counts_human_HGNC_symbols.tsv"), show_col_types = F)
meta <- read_tsv(here("testing", "data", "GSE72737_preprocessed_metadata.tsv"), show_col_types = F)
```

Preview the first few lines of input data. The `normalize_and_run_GSEA()` requires that the input expression data's first column is named "hgnc_symbol".

```{r render=lemon_print}
head(expr)
```

Preview the first few lines of the metadata. The `normalize_and_run_GSEA()` requires that the input group metadata's columns are named "sample_id", "group".

```{r render=lemon_print}
head(meta)
```

The samples should be in the same order in the expression data and metadata.

```{r}
stopifnot(identical(colnames(expr[,2:ncol(expr)]), meta$sample_id, attrib.as.set = F))
```

# Run GSEA

```{r eval=FALSE}
# note: this was run once and then set eval=FALSE to speed up runtime of Rmd
normalize_and_run_GSEA(expr_data = expr, 
                       groups = meta, 
                       control = "DNMT3A_WT", 
                       permute_mode = "gene_set", 
                       out_name = "DNMT3Amut", 
                       out_dir = here("testing", "results"), 
                       GSEAforR = "/Users/jcrockett/Documents/jobs/code_sample_projects/GSEAforR/")
```

# Load GSEA results

```{r}
result <- load_GSEA_results(here("testing", "results", "DNMT3Amut_perm_gene_set_h.all.v2023.2.Gsea.1710129235807"))
```

# Plot GSEA results

```{r GSEA_results_DNMT3A_Mut_vs_WT}
plot_GSEA_results(result = result, 
                  fdr_q_cutoff = 0.05, 
                  control_group = "DNMT3A WT", 
                  experimental_group = "DNMT3A Mutant", 
                  order = -1, 
                  save_file = here("testing", "results", "GSEA_results_DNMT3A_Mut_vs_WT.png"))
```



